server/schedulers/transfer_leader.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule"
+	"github.com/tikv/pd/server/schedule/filter"
+	"github.com/tikv/pd/server/schedule/operator"
+	"github.com/tikv/pd/server/schedule/plan"
+	"github.com/tikv/pd/server/storage/endpoint"
+)
+
+const (
+	// TransferLeaderName is transfer leader scheduler name.
+	TransferLeaderName = "transfer-leader-scheduler"
+	// TransferLeaderType is transfer leader scheduler type.
+	TransferLeaderType = "transfer-leader"
+	// TransferLeaderBatchSize is the number of operators to to transfer
+	// leaders by one scheduling
+	TransferLeaderBatchSize = 3
+	// TransferLeaderRecvMaxRegionSize is the max number of region can receive
+	// TODO: make it a reasonable value
+	TransferLeaderRecvMaxRegionSize = 1000
+)
+
+func init() {
+	schedule.RegisterSliceDecoderBuilder(TransferLeaderType, func(args []string) schedule.ConfigDecoder {
+		return func(v interface{}) error {
+			return nil
+		}
+	})
+
+	schedule.RegisterScheduler(TransferLeaderType, func(opController *schedule.OperatorController, _ endpoint.ConfigStorage, _ schedule.ConfigDecoder) (schedule.Scheduler, error) {
+		return newTransferLeaderScheduler(opController), nil
+	})
+}
+
+type transferLeaderScheduler struct {
+	*BaseScheduler
+	regions chan *core.RegionInfo
+}
+
+// newTransferLeaderScheduler creates an admin scheduler that transfers leader of a region.
+func newTransferLeaderScheduler(opController *schedule.OperatorController) schedule.Scheduler {
+	return &transferLeaderScheduler{
+		BaseScheduler: NewBaseScheduler(opController),
+		regions:       make(chan *core.RegionInfo, TransferLeaderRecvMaxRegionSize),
+	}
+}
+
+func (s *transferLeaderScheduler) GetName() string {
+	return TransferLeaderName
+}
+
+func (s *transferLeaderScheduler) GetType() string {
+	return TransferLeaderType
+}
+
+func (s *transferLeaderScheduler) IsScheduleAllowed(cluster schedule.Cluster) bool {
+	allowed := s.OpController.OperatorCount(operator.OpLeader) < cluster.GetOpts().GetLeaderScheduleLimit()
```

Comment:
Will it need to be restricted by the leader schedule limit?

Reply:
I'm not sure about this place. Normally, when this scheduler is needed, it's dealing with a failure case scenario, so it's important and urgent but maybe not common. What do you think this limit should be set to? (Or just construct the scenario test later to get a reasonable limit)